redis/cluster/redis.go

```diff
@@ -0,0 +1,182 @@
+	EnableTracing bool
+}
+
+// NewClient creates and returns a new redis client based on the given settings. It only supports redis 7 engine and uses go-redis v9.
+func NewClient(args *ClientArgs) (*Client, error) {
+	if args == nil {
+		return nil, fmt.Errorf("NewClient must have a non-nil args")
+	}
+
+	client := &Client{
+		ClusterMode: args.ClusterMode,
+	}
+
+	if args.ClusterMode {
+		if err := client.ConnectCluster(args.Url); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := client.Connect(args.Url); err != nil {
+			return nil, err
+		}
+	}
+
+	if args.EnableTracing {
+		if err := redisotel.InstrumentTracing(client.Instance); err != nil {
+			return nil, err
+		}
+	}
+
+	if args.EnableMetrics {
+		if err := redisotel.InstrumentMetrics(client.Instance); err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+// NewClientFromConfig creates and returns a new redis client based on the given settings from predefined env vars. It only supports redis 7 engine and uses go-redis v9.
+func NewClientFromConfig(config *viper.Viper, prefix string) (*Client, error) {
+	if config == nil {
+		return nil, fmt.Errorf("NewClientFromConfig must have a non-nil config")
+	}
+
+	args := CreateClientArgs(config, prefix)
+
+	return NewClient(args)
+}
+
+func CreateClientArgs(config *viper.Viper, prefix string) *ClientArgs {
+	url := config.GetString(fmt.Sprintf("%s.url", prefix))
+
+	if url == "" {
+		tls := config.GetBool(fmt.Sprintf("%s.tls", prefix))
+
+		endpoint := config.GetString(fmt.Sprintf("%s.endpoint", prefix))
+		port := config.GetString(fmt.Sprintf("%s.port", prefix))
+		pass := config.GetString(fmt.Sprintf("%s.password", prefix))
+
+		var baseUrl string
+		if tls {
+			baseUrl = "rediss://"
+		} else {
+			baseUrl = "redis://"
+		}
+
+		url = fmt.Sprintf("%s:%s@%s:%s", baseUrl, pass, endpoint, port)
+	}
+
+	clusterMode := config.GetBool(fmt.Sprintf("%s.clusterMode", prefix))
+	enableMetrics := config.GetBool(fmt.Sprintf("%s.enableMetrics", prefix))
+	enableTracing := config.GetBool(fmt.Sprintf("%s.enableTracing", prefix))
+
+	// Extra parameters
+	var queryParameters []string
+	maxRedirects := config.GetString(fmt.Sprintf("%s.maxRedirects", prefix))
+	if maxRedirects != "" {
+		queryParameters = append(queryParameters, fmt.Sprintf("max_redirects=%s", maxRedirects))
+	}
+
+	poolSize := config.GetString(fmt.Sprintf("%s.poolSize", prefix))
+	if poolSize != "" {
+		queryParameters = append(queryParameters, fmt.Sprintf("pool_size=%s", poolSize))
+	}
+
+	readTimeout := config.GetString(fmt.Sprintf("%s.timeout", prefix))
+	if readTimeout != "" {
+		queryParameters = append(queryParameters, fmt.Sprintf("read_timeout=%s", readTimeout))
+	}
+
+	if len(queryParameters) > 0 {
+		url = fmt.Sprintf("%s?%s", url, strings.Join(queryParameters, "&"))
+	}
+
+	return &ClientArgs{
+		ClusterMode:   clusterMode,
+		Url:           url,
+		EnableMetrics: enableMetrics,
+		EnableTracing: enableTracing,
+	}
+}
+
+// Connect to Redis
```

Comment:
should this function be exported? since it's being invoked when creating the client instance, not sure if allowing users to overwrite the internal universal redis instance is something desirable

same comment applies to `ConnectCluster`